packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart

```diff
@@ -0,0 +1,418 @@
+import 'flutter_icon_renderer.dart';
+
+/// Group of buttons where each button toggles the state of a VMService
+/// extension.
+///
+/// Use this class any time you need to write UI code that exposes button(s) to
+/// control VMServiceExtension. This class handles error handling and lifecycle
+/// states for keeping state on the client and device consistent so you don't
+/// have to.
+class ServiceExtensionButtonGroup extends StatefulWidget {
+  const ServiceExtensionButtonGroup({
+    this.minIncludeLabelWidth,
+    @required this.extensions,
+  });
+
+  final double minIncludeLabelWidth;
+  final List<ToggleableServiceExtensionDescription> extensions;
+
+  @override
+  _ServiceExtensionButtonGroupState createState() =>
+      _ServiceExtensionButtonGroupState();
+}
+
+/// Data class tracking the state of a single service extension.
+class ExtensionState {
+  ExtensionState(this.description);
+
+  final ToggleableServiceExtensionDescription description;
+  bool isSelected = false;
+  bool isAvailable = false;
+  StreamSubscription subscription;
+
+  void dispose() {
+    subscription?.cancel();
+    subscription = null;
+  }
+}
+
+class _ServiceExtensionButtonGroupState
+    extends State<ServiceExtensionButtonGroup> {
+  List<ExtensionState> _extensionStates;
+
+  @override
+  void initState() {
+    super.initState();
+    // To use ToggleButtons we have to track states for all buttons in the
+    // group here rather than tracking state with the individual button widgets
+    // which would be more natural.
+
+    _extensionStates = [for (var e in widget.extensions) ExtensionState(e)];
+
+    for (var extension in _extensionStates) {
+      // Listen for changes to the state of each service extension using the
+      // VMServiceManager.
+      final extensionName = extension.description.extension;
+      // Update the button state to match the latest state on the VM.
+      serviceManager.serviceExtensionManager
+          .getServiceExtensionState(extensionName, (state) {
+        setState(() {
+          extension.isSelected =
+              state.value == extension.description.enabledValue;
+        });
+      });
+      // Track whether the extension is actually exposed by the VM.
+      extension.isAvailable = serviceManager.serviceExtensionManager
+          .isServiceExtensionAvailable(extensionName);
+      extension.subscription =
+          serviceManager.serviceExtensionManager.hasServiceExtension(
+        extensionName,
+        (available) {
+          setState(() {
+            extension.isAvailable = available;
+          });
+        },
+      );
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    for (var config in _extensionStates) {
+      config.dispose();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    // TODO(jacobr): respect _available better by displaying whether individual
+    // widgets are available (not currently supported by ToggleButtons).
```

Comment:
We might be better off using multiple regular Material Buttons and customizing their appearance. 

That would save us the trouble of working with a widget that makes assumptions about its underlying data model.